File: SignUpLogin/SignUpLogin/Controllers/LoginController.cs
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using SignUpLogin.Data;
using SignUpLogin.Models;
using SignUpLogin.Models.ViewModels;

namespace SignUpLogin.Controllers
{
    public class LoginController : Controller
    {
        private readonly ApplicationDbContext _context;
        private const string RememberedIdCookie = "RememberedIdNumber";

        public LoginController(ApplicationDbContext context)
        {
            _context = context;
        }

        public async Task<IActionResult> Index()
        {
            var model = new Login();

            // Handle Remember Me Cookie
            if (Request.Cookies.TryGetValue(RememberedIdCookie, out var rememberedIdNumber) && !string.IsNullOrWhiteSpace(rememberedIdNumber))
            {
                model.IdNumber = rememberedIdNumber;
                model.RememberMe = true;
            }

            // Fetch Top 5 Students for Leaderboard
            var topStudents = await _context.StudentPoints
                .AsNoTracking()
                .Include(p => p.Student)
                .OrderByDescending(p => p.Points)
                .Take(5)
                .Select(p => new LeaderboardEntryViewModel
                {
                    StudentIdNumber = p.StudentIdNumber,
                    StudentName = p.Student != null ? $"{p.Student.FirstName} {p.Student.LastName}".Trim() : p.StudentIdNumber,
                    Points = p.Points,
                    LifetimePoints = p.LifetimePoints
                })
                .ToListAsync();

            ViewBag.LeaderboardData = topStudents;

            return View("Login", model);
        }

        [HttpPost]
        public async Task<IActionResult> Index(Login model)
        {
            if (!ModelState.IsValid)
            {
                // Re-fetch leaderboard if validation fails so it still shows
                ViewBag.LeaderboardData = await GetLeaderboardData();
                return View("Login", model);
            }

            // Try by IdNumber first (students), then by FirstName for admins
            var user = await _context.Signups
                           .FirstOrDefaultAsync(u => u.IdNumber == model.IdNumber)
                       ?? await _context.Signups
                           .FirstOrDefaultAsync(u => u.FirstName == model.IdNumber && u.Role == "Admin");

            if (user == null || !BCrypt.Net.BCrypt.Verify(model.Password, user.Password))
            {
                ModelState.AddModelError(string.Empty, "Invalid ID Number/Username or Password.");
                // Re-fetch leaderboard if login fails
                ViewBag.LeaderboardData = await GetLeaderboardData();
                return View("Login", model);
            }

            HttpContext.Session.Clear();

            if (model.RememberMe)
            {
                Response.Cookies.Append(RememberedIdCookie, model.IdNumber, new CookieOptions
                {
                    Expires = DateTimeOffset.UtcNow.AddDays(30),
                    HttpOnly = true,
                    IsEssential = true,
                    Secure = Request.IsHttps,
                    SameSite = SameSiteMode.Lax
                });
            }
            else
            {
                Response.Cookies.Delete(RememberedIdCookie);
            }

            HttpContext.Session.SetString("UserName", $"{user.FirstName} {user.LastName}");
            HttpContext.Session.SetString("IdNumber", user.IdNumber);
            HttpContext.Session.SetString("Course", user.Course);
            HttpContext.Session.SetString("CourseLevel", user.CourseLevel);
            HttpContext.Session.SetString("Email", user.Email);
            HttpContext.Session.SetString("Role", user.Role);
            HttpContext.Session.SetString("ProfileImagePath", user.ProfileImagePath ?? string.Empty);

            TempData["LoginSuccess"] = $"Welcome back, {user.FirstName}!";

            if (user.Role == "Admin")
            {
                return RedirectToAction("Home", "Admin");
            }
            else
            {
                return RedirectToAction("Index", "Home");
            }
        }

        public IActionResult Logout()
        {
            HttpContext.Session.Clear();
            TempData["LoggedOut"] = "You have been logged out successfully.";
            return RedirectToAction("Index");
        }

        // Helper method to avoid code duplication
        private async Task<List<LeaderboardEntryViewModel>> GetLeaderboardData()
        {
            return await _context.StudentPoints
                .AsNoTracking()
                .Include(p => p.Student)
                .OrderByDescending(p => p.Points)
                .Take(5)
                .Select(p => new LeaderboardEntryViewModel
                {
                    StudentIdNumber = p.StudentIdNumber,
                    StudentName = p.Student != null ? $"{p.Student.FirstName} {p.Student.LastName}".Trim() : p.StudentIdNumber,
                    Points = p.Points,
                    LifetimePoints = p.LifetimePoints
                })
                .ToListAsync();
        }
    }
}

File: SignUpLogin_PHP/controllers/LoginController.php
<?php
// controllers/LoginController.php

require_once __DIR__ . '/../config/config.php';
require_once __DIR__ . '/../config/database.php';
require_once __DIR__ . '/../models/Signup.php';

class LoginController {
    private $db;
    private $signupModel;

    public function __construct() {
        $database = new Database();
        $this->db = $database->getConnection();
        $this->signupModel = new Signup($this->db);
    }

    public function index() {
        // Check if already logged in
        if (isLoggedIn()) {
            if (isAdmin()) {
                redirect('admin/home');
            } else {
                redirect('home');
            }
        }

        // Get remembered ID from cookie
        $rememberedIdNumber = '';
        if (isset($_COOKIE['RememberedIdNumber'])) {
            $rememberedIdNumber = $_COOKIE['RememberedIdNumber'];
        }

        include __DIR__ . '/../views/login/login.php';
    }

    public function login() {
        if ($_SERVER['REQUEST_METHOD'] === 'POST') {
            $idNumber = isset($_POST['IdNumber']) ? trim($_POST['IdNumber']) : '';
            $password = isset($_POST['Password']) ? $_POST['Password'] : '';
            $rememberMe = isset($_POST['RememberMe']);

            $errors = [];

            // Validation
            if (empty($idNumber)) {
                $errors[] = "ID Number is required.";
            }
            if (empty($password)) {
                $errors[] = "Password is required.";
            }

            if (empty($errors)) {
                // Try by IdNumber first (students), then by FirstName for admins
                $user = $this->signupModel->getByIdNumber($idNumber);
                
                if (!$user) {
                    $user = $this->signupModel->getByNameAndRole($idNumber, 'Admin');
                }

                if (!$user || !password_verify($password, $user['Password'])) {
                    $errors[] = "Invalid ID Number/Username or Password.";
                } else {
                    // Set session variables
                    $_SESSION['UserName'] = $user['FirstName'] . ' ' . $user['LastName'];
                    $_SESSION['IdNumber'] = $user['IdNumber'];
                    $_SESSION['Course'] = $user['Course'];
                    $_SESSION['CourseLevel'] = $user['CourseLevel'];
                    $_SESSION['Email'] = $user['Email'];
                    $_SESSION['Role'] = $user['Role'];
                    $_SESSION['ProfileImagePath'] = $user['ProfileImagePath'] ?? '';

                    // Handle remember me
                    if ($rememberMe) {
                        setcookie('RememberedIdNumber', $idNumber, time() + (30 * 24 * 60 * 60), '/', '', false, true);
                    } else {
                        setcookie('RememberedIdNumber', '', time() - 3600, '/');
                    }

                    // Redirect based on role
                    if ($user['Role'] === 'Admin') {
                        redirect('admin/home');
                    } else {
                        redirect('home');
                    }
                }
            }

            // If we have errors, show the login form again
            $rememberedIdNumber = $idNumber;
            include __DIR__ . '/../views/login/login.php';
        } else {
            $this->index();
        }
    }

    public function logout() {
        // Clear session
        $_SESSION = array();
        session_destroy();

        // Set logout message
        $_SESSION['LoggedOut'] = "You have been logged out successfully.";

        // Redirect to login
        redirect('login');
    }
}
?>
